core/src/androidMain/kotlin/Connection.kt

```diff
@@ -52,12 +58,25 @@ internal class Connection(
     suspend inline fun <reified T> execute(
         crossinline action: BluetoothGatt.() -> Boolean,
     ): T = lock.withLock {
+        if (pending) {
+            // Discard response as we've performed another `execute` without the previous finishing. This happens if a
+            // previous `execute` was cancelled after invoking GATT action, but before receiving response from callback
+            // channel. See https://github.com/JuulLabs/kable/issues/326 for more details.
+            val response = callback.onResponse.receive()
+            pending = false
+            logger.warn {
+                message = "Discarded response"
+                detail("response", response.toString())
+            }
+        }
+
         withContext(dispatcher) {
+            pending = true
             action.invoke(bluetoothGatt) || throw GattRequestRejectedException()
         }
 
         val response = try {
-            callback.onResponse.receive()
+            callback.onResponse.receive().also { pending = false }
```

Comment:
Should this be a `finally`?

Reply:
No, as I only want to "reset" the `pending` state if we successfully received a response.
If the Coroutine was cancelled then I want `pending` to remain `true` for the next pass.

I wasn't excited about carrying state across multiple request/response operations, but there isn't a way to clear a `Channel` (that I am aware of); which is what I would've done in a `catch (e: CancellationException)` block if that were possible.